File: algorithms/python/lcs.py
import sys
import time

def lcs(a, b):
    n, m = len(a), len(b)
    prev = [0] * (m + 1)
    curr = [0] * (m + 1)

    for i in range(1, n + 1):
        ai = a[i-1]
        for j in range(1, m + 1):
            if ai == b[j-1]:
                curr[j] = prev[j-1] + 1
            else:
                curr[j] = curr[j-1] if curr[j-1] > prev[j] else prev[j]
        prev, curr = curr, prev
        for j in range(m + 1): curr[j] = 0

    return prev[m]

def main():
    data   = sys.stdin.read().splitlines()
    a      = data[0].strip()
    b      = data[1].strip()

    start  = time.perf_counter()
    result = lcs(a, b)
    ms     = (time.perf_counter() - start) * 1000

    sys.stdout.write(f"{result}\n")
    sys.stderr.write(f"{ms:.3f}\n")

if __name__ == "__main__":
    main()

File: algorithms/lua/lcs.lua
local function lcs(a, b)
    local n, m = #a, #b
    local prev = {}
    local curr = {}
    for j = 0, m do prev[j] = 0; curr[j] = 0 end

    for i = 1, n do
        local ai = a:sub(i, i)
        for j = 1, m do
            if ai == b:sub(j, j) then
                curr[j] = prev[j-1] + 1
            else
                curr[j] = curr[j-1] > prev[j] and curr[j-1] or prev[j]
            end
        end
        prev, curr = curr, prev
        for j = 0, m do curr[j] = 0 end
    end
    return prev[m]
end

local function main()
    local a   = io.read("*l")
    local b   = io.read("*l")

    local start  = os.clock()
    local result = lcs(a, b)
    local ms     = (os.clock() - start) * 1000

    print(result)
    io.stderr:write(string.format("%.3f\n", ms))
end

main()